### PR TITLE
Move CI from GitHub Actions to GitLab

### PR DESCRIPTION
**Pull request by TriplEight**
# New CI.

### Nightly test pipeline
Is triggered by a CI image rebuild at 6 AM CET.
It gets a new Rust nightly and runs all types of stable and nightly tests.
Doesn't run builds.
Looks like this:
image
(but upstream will be the scripts project)

### Nightly build pipeline
Is triggered by a schedule at 1 AM CET.
It runs all the regular tests and then builds and publishing.
Looks like this:
image

### New tag pipeline
Is triggered by a push of a new git tag.
Similar to the Nightly build pipeline.

### Master pipeline
Is triggered by a new commit in the `master` branch.
Looks like this:
image

### PR pipeline
Is triggered by a new commit in any PR.
For now similar to the Master pipeline.

### Web pipeline
Is triggered by [Run Pipeline](https://gitlab.parity.io/parity/parity-bridges-common/-/pipelines) button.
Depends on which branch is set there, generally similar to PR or Master pipeline.

### CODE REVIEW

File: .gitlab-ci.yml
Comment (tomusdrw) [inline .gitlab-ci.yml:62]: This is not really needed, we don't care about WASM.

File: .gitlab-ci.yml
Comment (tomusdrw) [inline .gitlab-ci.yml:82]: same here.

**Review by tomusdrw — COMMENTED** [review]
Nice! We still need `fmt`, `clippy` and `cargo check` and docker publish (images nightly) though.

**Comment by HCastano** [comment]
Any updates here @TriplEight?

**Comment by TriplEight** [comment]
New pipeline [example](https://gitlab.parity.io/parity/parity-bridges-common/-/pipelines/138854).
Please highlight when do you want which checks, whether should all of them run on PRs, on tags, on master and on schedules?
Do you want `fmt` on stable, because that's how it was in GHA.
Do you want artifacts from both stable and nightly build jobs?
When do you want to publish an image?
Which jobs you want to be non-blocking?

**Comment by HCastano** [comment]
> New pipeline example.

👌

> Please highlight when do you want which checks, whether should all of them run on PRs, on tags, on master and on schedules?

So I don't think we need to run the `nightly` compiler checks on every PR. If possible I think we should run those overnight alongside the image/artifact build.

As for the timing we'll want to run all the checks (except the `nightly` ones) on every single PR, and it would also be good to run them on `master` when a new PR is merged. 

Tags are a little special, and we haven't been using them yet tbh. However, we should run the full CI suite + image/artifact jobs when a new tag is pushed.

> Do you want fmt on stable, because that's how it was in GHA?

Yes, please run `RustFmt` on stable

> Do you want artifacts from both stable and nightly build jobs?

No, only artifacts built with the `stable` compiler should be published.

> When do you want to publish an image?

Images should be built from the `master` branch nightly (midnight UTC). Binaries should also be published alongside this image

> Which jobs you want to be non-blocking?

We don't really care if `cargo-deny` fails for security vuls since most of the time we can't do anything about them. However, we do care if it fails with a license check. If that logic is too hard, just mark it as failed and we can manually check on the reasoning and deal with it accordingly.

Everything else, including Clippy and the Spellcheck, should be required

Thanks for looking at this again @TriplEight

**Comment by TriplEight** [comment]
Thanks

> So I don't think we need to run the `nightly` compiler checks on every PR. If possible I think we should run those overnight alongside the image/artifact build.

Then since you already have `nightly` rust installed in the CI image, I'll run the `nightly` tests on every CI image (hence `nightly` Rust update). You will see these tests in the same Gitlab pipelines.

Other questions about CI:
- it's contradictive: either you want **all** nightly-rust stuff running on schedules or `clippy` which is also nightly on all PR commits
- why do you even need `nightly` `build` when you have both `nightly` `test` and `check`
- can I optimize and run `build` together with `test` but making them both `--release`?

**Comment by HCastano** [comment]
>  it's contradictive: either you want **all** nightly-rust stuff running on schedules or `clippy` which is also nightly on all PR commits

I wouldn't quite say that. Just because we have Clippy running on every PR doesn't mean we need to have the rest of the `nightly` dependent checks running on every PR. We need the `nightly` compiler for the `stable` PR builds anyways because of WASM

>  why do you even need `nightly` `build` when you have both `nightly` `test` and `check`

So I think we should run nightly builds using the `nightly` compiler which do both `build` and `test`. We want to do this to be warned of any potential breakages coming our way (or to report them to the compiler team)

>  can I optimize and run `build` together with `test` but making them both `--release`?

What do you mean by "run them together"?

**Comment by TriplEight** [comment]
> What do you mean by "run them together"?

just to reuse the cache within the same job. Given the same flags, it's often quite fast (sometimes seconds) to run `cargo build x` and then `cargo test x`. And then, if tests are good, the binary from build will be passed further.

**Comment by TriplEight** [comment]
I've executed a [bench](https://gitlab.parity.io/TriplEight/parity-bridges-common/-/pipelines/140370/builds): build + test in one job instead of both in separate jobs, no cache.
Here, `cargo test` runs with `--release` in both cases:
image

Here a separate test job runs without `--release`
image

## Results

In case of build pipeline it's almost free to execute test together with build, but we do not intend to build often, so I'll stick to having them separately and `cargo test` without the `--release` flag, this will be the fastest.

**Comment by TriplEight** [comment]
This is ready to be merged.
I'll take care of publishing binaries to s3 in the next PR, can't do it without CI merged to master.

**Comment by TriplEight** [comment]
Before the merge, I'll switch the required tests by github to the ones that are coming from gitlab

File: .gitlab-ci.yml
Comment (tomusdrw) [inline .gitlab-ci.yml:258]: Don't we need to update `docker-compose.yml`?

**Review by tomusdrw — APPROVED** [review]
Awesome, thanks!

File: .config/spellcheck.toml
Comment (TriplEight) [reply .config/spellcheck.toml:6]: I've contacted spellcheck's maintainer on why it didn't use the built-in dictionary and they told me this way to activate it.

File: .gitlab-ci.yml
Comment (TriplEight) [reply .gitlab-ci.yml:18]: [This](https://github.com/paritytech/scripts/blob/master/.gitlab-ci.yml#L375-L384) is how the image updates.
image
1. the new CI image is built at 6 AM nightly and pushed with `staging` tag.
2. `staging` is tested against `bridges` CI on master and gets a vuln scan
3. if green, it gets into production

in this case it's failed because Gitlab CI is not in master yet.

File: .gitlab-ci.yml
Comment (TriplEight) [reply .gitlab-ci.yml:258]: I need to know more about it

**Comment by HCastano** [comment]
I didn't get a chance to review this!

File: modules/grandpa/src/lib.rs
Comment (HCastano) [inline modules/grandpa/src/lib.rs:696]: Tbh I'm not a fan of this Clippy suggestion, the other was easier to read (no double !!), but 🤷‍♂️

File: ci.Dockerfile
Comment (HCastano) [inline ci.Dockerfile:1]: To clarify, what's going on here is that we're taking the binaries built by the CI pipeline and putting them inside a Dockerfile, whose image we then publish?

Where does this leave our top level Dockerfile then?

**Review by HCastano — COMMENTED** [review]
In general the pipeline looks good, I still have some questions though

File: .gitlab-ci.yml
Comment (TriplEight) [reply .gitlab-ci.yml:198]: It's built anyway with `--workspace`. Do you mean we shouldn't use this artifact for now? Ain't it used on one of the Dockerfiles?

File: .gitlab-ci.yml
Comment (TriplEight) [reply .gitlab-ci.yml:240]: Oh missed it here, thanks!

File: Dockerfile
Comment (TriplEight) [reply Dockerfile:18]: every `RUN ...` introduces a new layer in the image. It's good when you're debugging builds, but adds additional weight to the resulting image.

File: README.md
Comment (TriplEight) [reply README.md:52]: ah ok, was satisfying the md linter xD

File: ci.Dockerfile
Comment (TriplEight) [reply ci.Dockerfile:1]: Yeah, exactly.
Top-level `./Dockerfile` is a CI on its own with that docker builder pattern. It's not beneficial to use such a pattern within another CI system as it won't reuse cache normally.
However, it's great for local usage.
I couldn't come up with the idea on how to inherit one from another to stay DRY, there's a `--target` in `docker build` which can choose the exact step of the builder pattern to start from, but
- the unavoidable `COPY` instruction will trigger the initial step anyway, hence it won't help
- it's not yet working in buldah man_shrugging

File: modules/grandpa/src/lib.rs
Comment (tomusdrw) [reply modules/grandpa/src/lib.rs:696]: It gives much better debugging info if it fails (shows the entire expression instead of `expected true, got false`).

File: .gitlab-ci.yml
Comment (tomusdrw) [reply .gitlab-ci.yml:198]: Let's keep them until they are part of the codebase I don't think it's going to affect build times that much.

File: .gitlab-ci.yml
Comment (tomusdrw) [reply .gitlab-ci.yml:258]: I thought that the docker image names might have changed and the docker-compose file is fetching images from dockerhub, but I guess there were no changes.

File: .gitlab-ci.yml
Comment (TriplEight) [reply .gitlab-ci.yml:198]: as long as we run `cargo build --workspace` we're simply building everything

File: ci.Dockerfile
Comment (HCastano) [reply ci.Dockerfile:1]: I think we'll need to note down somewhere that our published images don't match that of the `Dockerfile`. I'm afraid it could cause some confusion if people use different images without realizing

**Pull request merged**
